src/auth.zig

```diff
@@ -5,23 +5,137 @@ const std = @import("std");
 const Allocator = std.mem.Allocator;
 
 pub const Auth = @This();
+pub const User = @import("auth/user.zig");
 
-user: User = User{},
+provider: AnyAuth,
+current_user: ?User = null,
 
-pub const User = struct {
-    username: []const u8 = "invalid username",
+pub const Error = error{
+    UnknownUser,
+    Unauthenticated,
+    NotProvided,
 };
 
 pub fn valid(a: Auth) bool {
-    _ = a;
-    return true;
+    return a.provider.valid() catch false;
 }
 
-pub fn validOrError(a: Auth) !void {
-    if (!a.valid()) return error.Unauthenticated;
+pub fn requireValid(a: Auth) !void {
+    if (a.current_user == null or !a.valid()) return error.Unauthenticated;
 }
 
-pub fn currentUser(a: Auth, alloc: Allocator) !User {
-    _ = alloc;
-    return a.user;
+pub fn Provider(T: type) type {
+    return struct {
+        const Self = @This();
+        ctx: T,
```

Comment:
Whats the point of this variable?

Reply:
I had the same question my self, and almost deleted it. But I assume there will be some auth provider that will want to be stateful. This is for that. But the few that I can think of aren't, so I'm torn one including it or not. It might be better to delete it until we need it (but that does preclude anything downstream from having a state before we introduce one)

All that said, I'm still torn, and happy to drop it until then.

Reply:
We can leave it for now.